rioxarray/raster_index.py

```diff
@@ -0,0 +1,428 @@
+        labels = np.asarray(coord_labels[self.coord_name])
+
+        if self.is_xaxis:
+            positions, _ = ~self.affine * (labels, np.zeros_like(labels))
+        else:
+            _, positions = ~self.affine * (np.zeros_like(labels), labels)
+
+        return {self.dim: positions}
+
+    def equals(self, other):
+        if not isinstance(other, AxisAffineTransform):
+            return False
+
+        # only compare the affine parameters of the relevant axis
+        if self.is_xaxis:
+            affine_match = (
+                self.affine.a == other.affine.a and self.affine.c == other.affine.c
+            )
+        else:
+            affine_match = (
+                self.affine.e == other.affine.e and self.affine.f == other.affine.f
+            )
+
+        return affine_match and self.size == other.size
+
+    def generate_coords(
+        self, dims: tuple[str, ...] | None = None
+    ) -> dict[Hashable, Any]:
+        assert dims is None or dims == self.dims
+        return self.forward({self.dim: np.arange(self.size)})
+
+    def slice(self, slice: slice) -> "AxisAffineTransform":
+        start = max(slice.start or 0, 0)
+        stop = min(slice.stop or self.size, self.size)
+        step = slice.step or 1
+
+        # TODO: support reverse transform (i.e., start > stop)?
+        assert start < stop
+
+        size = (stop - start) // step
+        scale = float(step)
+
+        if self.is_xaxis:
+            affine = (
+                self.affine * Affine.translation(start, 0.0) * Affine.scale(scale, 1.0)
+            )
+        else:
+            affine = (
+                self.affine * Affine.translation(0.0, start) * Affine.scale(1.0, scale)
+            )
+
+        return type(self)(
+            affine,
+            size,
+            self.coord_name,
+            self.dim,
+            is_xaxis=self.is_xaxis,
+            dtype=self.dtype,
+        )
+
+
+class AxisAffineTransformIndex(CoordinateTransformIndex):
+    """Axis-independent Xarray Index for an affine 2D transform with no
+    skew/rotation.
+
+    For internal use only.
+
+    This Index class provides specific behavior on top of
+    Xarray's `CoordinateTransformIndex`:
+
+    - Data slicing computes a new affine transform and returns a new
+      `AxisAffineTransformIndex` object
+
+    - Otherwise data selection creates and returns a new Xarray
+      `PandasIndex` object for non-scalar indexers
+
+    - The index can be converted to a `pandas.Index` object (useful for Xarray
+      operations that don't work with Xarray indexes yet).
+
+    """
+
+    axis_transform: AxisAffineTransform
+    dim: str
+
+    def __init__(self, transform: AxisAffineTransform):
+        assert isinstance(transform, AxisAffineTransform)
+        super().__init__(transform)
+        self.axis_transform = transform
+        self.dim = transform.dim
+
+    def isel(  # type: ignore[override]
```

Comment:
Again here we should make the return type of `xarray.core.Index.isel()` more permissive (i.e., any `Index` object or `None`) so we can remove the `# type: ignore` comment.